pytest 4.0.0 causes errors

pytest is installed in .tox/py27/.../site-packages, not in moin-venv-python. A workaround was added to requirements.d/development.txt:

        tox
        pytest<4.0.0
        pytest-pep8
        # we use lxml.etree for xpath-based testing
        lxml

Partial test output with pytest 4.0.0 installed:

         ERROR collecting .tox/py27/lib/site-packages/moin/storage/stores/_tests/test_fs.py 
        .tox\py27\lib\site-packages\pluggy\hooks.py:284: in __call__
            return self._hookexec(self, self.get_hookimpls(), kwargs)
        .tox\py27\lib\site-packages\pluggy\manager.py:67: in _hookexec
            return self._inner_hookexec(hook, methods, kwargs)
        .tox\py27\lib\site-packages\pluggy\manager.py:61: in <lambda>
            firstresult=hook.spec.opts.get("firstresult") if hook.spec else False,
        .tox\py27\lib\site-packages\_pytest\python.py:224: in pytest_pycollect_makeitem
            res = list(collector._genfunctions(name, obj))
        .tox\py27\lib\site-packages\_pytest\python.py:409: in _genfunctions
            self.ihook.pytest_generate_tests(metafunc=metafunc)
        .tox\py27\lib\site-packages\pluggy\hooks.py:284: in __call__
            return self._hookexec(self, self.get_hookimpls(), kwargs)
        .tox\py27\lib\site-packages\pluggy\manager.py:67: in _hookexec
            return self._inner_hookexec(hook, methods, kwargs)
        .tox\py27\lib\site-packages\pluggy\manager.py:61: in <lambda>
            firstresult=hook.spec.opts.get("firstresult") if hook.spec else False,
        .tox\py27\lib\site-packages\moin\storage\stores\_tests\conftest.py:60: in pytest_generate_tests
            stores = multi_mark.kwargs['Store']
        .tox\py27\lib\site-packages\_pytest\mark\structures.py:25: in warned
            warnings.warn(warning, stacklevel=2)
        E   RemovedInPytest4Warning: MarkInfo objects are deprecated as they contain merged marks which are hard to deal with correctly.
        E   Please use node.get_closest_marker(name) or node.iter_markers(name).
        E   Docs: https://docs.pytest.org/en/latest/mark.html#updating-code
         ERROR collecting .tox/py27/lib/site-packages/moin/storage/stores/_tests/test_memory.py 
        .tox\py27\lib\site-packages\pluggy\hooks.py:284: in __call__
            return self._hookexec(self, self.get_hookimpls(), kwargs)
        .tox\py27\lib\site-packages\pluggy\manager.py:67: in _hookexec
            return self._inner_hookexec(hook, methods, kwargs)
        .tox\py27\lib\site-packages\pluggy\manager.py:61: in <lambda>
            firstresult=hook.spec.opts.get("firstresult") if hook.spec else False,
        .tox\py27\lib\site-packages\_pytest\python.py:224: in pytest_pycollect_makeitem
            res = list(collector._genfunctions(name, obj))
        .tox\py27\lib\site-packages\_pytest\python.py:409: in _genfunctions
            self.ihook.pytest_generate_tests(metafunc=metafunc)
        .tox\py27\lib\site-packages\pluggy\hooks.py:284: in __call__
            return self._hookexec(self, self.get_hookimpls(), kwargs)
        .tox\py27\lib\site-packages\pluggy\manager.py:67: in _hookexec
            return self._inner_hookexec(hook, methods, kwargs)
        .tox\py27\lib\site-packages\pluggy\manager.py:61: in <lambda>
            firstresult=hook.spec.opts.get("firstresult") if hook.spec else False,
        .tox\py27\lib\site-packages\moin\storage\stores\_tests\conftest.py:60: in pytest_generate_tests
            stores = multi_mark.kwargs['Store']
        .tox\py27\lib\site-packages\_pytest\mark\structures.py:25: in warned
            warnings.warn(warning, stacklevel=2)
        E   RemovedInPytest4Warning: MarkInfo objects are deprecated as they contain merged marks which are hard to deal with correctly.
        E   Please use node.get_closest_marker(name) or node.iter_markers(name).
        E   Docs: https://docs.pytest.org/en/latest/mark.html#updating-code
         ERROR collecting .tox/py27/lib/site-packages/moin/storage/stores/_tests/test_sqlite.py 
        .tox\py27\lib\site-packages\pluggy\hooks.py:284: in __call__
            return self._hookexec(self, self.get_hookimpls(), kwargs)
        .tox\py27\lib\site-packages\pluggy\manager.py:67: in _hookexec
            return self._inner_hookexec(hook, methods, kwargs)
        .tox\py27\lib\site-packages\pluggy\manager.py:61: in <lambda>
            firstresult=hook.spec.opts.get("firstresult") if hook.spec else False,
        .tox\py27\lib\site-packages\_pytest\python.py:224: in pytest_pycollect_makeitem
            res = list(collector._genfunctions(name, obj))
        .tox\py27\lib\site-packages\_pytest\python.py:409: in _genfunctions
            self.ihook.pytest_generate_tests(metafunc=metafunc)
        .tox\py27\lib\site-packages\pluggy\hooks.py:284: in __call__
            return self._hookexec(self, self.get_hookimpls(), kwargs)
        .tox\py27\lib\site-packages\pluggy\manager.py:67: in _hookexec
            return self._inner_hookexec(hook, methods, kwargs)
        .tox\py27\lib\site-packages\pluggy\manager.py:61: in <lambda>
            firstresult=hook.spec.opts.get("firstresult") if hook.spec else False,
        .tox\py27\lib\site-packages\moin\storage\stores\_tests\conftest.py:60: in pytest_generate_tests
            stores = multi_mark.kwargs['Store']
        .tox\py27\lib\site-packages\_pytest\mark\structures.py:25: in warned
            warnings.warn(warning, stacklevel=2)
        E   RemovedInPytest4Warning: MarkInfo objects are deprecated as they contain merged marks which are hard to deal with correctly.
        E   Please use node.get_closest_marker(name) or node.iter_markers(name).
        E   Docs: https://docs.pytest.org/en/latest/mark.html#updating-code

--- a/requirements.d/development.txt
+++ b/requirements.d/development.txt
@@ -1,5 +1,5 @@
 tox
-pytest<4.0.0
+pytest
 pytest-pep8
 # we use lxml.etree for xpath-based testing
 lxml

--- a/src/moin/storage/stores/_tests/conftest.py
+++ b/src/moin/storage/stores/_tests/conftest.py
@@ -52,7 +52,7 @@ def pytest_generate_tests(metafunc):
                 argvalues.append((storename, klass))
         metafunc.parametrize(argname, argvalues, ids=ids, indirect=True)
 
-    multi_mark = getattr(metafunc.function, 'multi', None)
+    multi_mark = metafunc.definition.get_closest_marker('multi')
     if multi_mark is not None:
         # XXX: hack
         ids = []

--- a/src/moin/storage/stores/_tests/conftest.py
+++ b/src/moin/storage/stores/_tests/conftest.py
@@ -28,7 +28,7 @@ constructors = {
 
 
 def pytest_generate_tests(metafunc):
-    argnames = metafunc.funcargnames
+    argnames = metafunc.fixturenames
 
     if 'store' in argnames:
         klasses = 'BytesStore', 'FileStore'

--- a/src/moin/storage/stores/_tests/conftest.py
+++ b/src/moin/storage/stores/_tests/conftest.py
@@ -54,7 +54,6 @@ def pytest_generate_tests(metafunc):
 
     multi_mark = metafunc.definition.get_closest_marker('multi')
     if multi_mark is not None:
-        # XXX: hack
         ids = []
         argvalues = []
         stores = multi_mark.kwargs['Store']

